sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -83,15 +83,18 @@ trait ShuffleExchangeLike extends Exchange {
 case class ShuffleExchangeExec(
     override val outputPartitioning: Partitioning,
     child: SparkPlan,
-    noUserSpecifiedNumPartition: Boolean = true) extends ShuffleExchangeLike {
+    noUserSpecifiedNumPartition: Boolean = true,
+    isNullAwareAntiJoin: Boolean = false) extends ShuffleExchangeLike {
```

Comment:
Just personal opinion - I feel NULL-aware anti join is not very popular query patterns, is there a way to avoid adding parameter here for popular class `ShuffleExchangeExec`? It's kind of weird to read that shuffle exchange depending on a config for NULL-aware anti join.

Reply:
It is popular somewhat: The NOT IN (Subquery) is basically a NAAJ. The results of the subquery can be huge and I think that's the case handled in this PR.

Reply:
sorry I think I should be talking with data in first place. I just checked and found in our case, less than 10% of join is using `BroadcastNestedLoopJoin` and amount of CPU is also single digit percentage. Our workload by no means is representative for industry, would like to hear more of based on data.

Reply:
> It is popular somewhat: The NOT IN (Subquery) is basically a NAAJ. The results of the subquery can be huge and I think that's the case handled in this PR.

Nice to have you here, ^_^

Reply:
> Just personal opinion - I feel NULL-aware anti join is not very popular query patterns, is there a way to avoid adding parameter here for popular class `ShuffleExchangeExec`? It's kind of weird to read that shuffle exchange depending on a config for NULL-aware anti join.

Yes. I found it weird to push extra parameter in ShuffleExchangeExec too. I've got two option in the beginning

* Count nullKey and put them into ShuffleExchangeExec SQLMetrics for all Query.
* Count nullKey only for NAAJ Query but add extra parameter.

Maybe you knew about a better way to identify nullKey in Exchange, please feel free to share your idea.

Reply:
> sorry I think I should be talking with data in first place. I just checked and found in our case, less than 10% of join is using `BroadcastNestedLoopJoin` and amount of CPU is also single digit percentage. Our workload by no means is representative for industry, would like to hear more of based on data.

First, some of the interest in improving single key NAAJ stems from TPCH-Q16: Benchmarketing. Without this optimization that one query can easily be a huge portion of the total TPCH runtime and skew the results.

Having said that, one of the chief reasons to optimize NOT-IN is in terms of avoiding customer surprises. Often, people will trip on this query after an innocuous change. This is particularly true when customers are submitting queries via a UI tool (where you don't have control over the query generated by the tool) or when they innocently introduce this query via a Notebook. 

It's not very interesting to optimize in the batch sense, because you have some flexibility to change the pipeline to use Not-Exists instead of Not-In (which is not planned as an NAAJ).

Reply:
Hi Devesh, I am totally with you for benchmarketing perspective and avoid customer surprise. I think we are discussing whether this kind of query pattern is popular or not. I think a renaming for the parameter of `isNullAwareAntiJoin` can alleviate my concern.

Reply:
This is too hacky, we shouldn't pollute the general shuffle node for one specific case.

I think we need to develop a flexible stats collection framework for AQE first. NAAJ is just one use case: it asks AQE to collect null count.

Reply:
@c21, Yeah @leanken proposed the name `collectNullPartitionKeyMetrics` below. I also made the similar observation below that we shouldn't have very join specific naming cropping up in shuffle code.